### PR TITLE
GODRIVER-1390 Add support for shorter SCRAM conversation

### PR DESCRIPTION
**Pull request by divjotarora**

### CODE REVIEW

File: x/mongo/driver/auth/plain.go [hunk 1]
```diff
@@ -42,6 +42,8 @@ type plainSaslClient struct {
 	password string
 }
 
+var _ SaslClient = (*plainSaslClient)(nil)
```
Comment (divjotarora) [inline x/mongo/driver/auth/plain.go]: Added type assertions for clarity.

File: x/mongo/driver/auth/sasl.go [hunk 2]
```diff
@@ -20,6 +20,7 @@ type SaslClient interface {
 	Start() (string, []byte, error)
 	Next(challenge []byte) ([]byte, error)
 	Completed() bool
+	ShorterConversationSupported() bool
```
Comment (divjotarora) [inline x/mongo/driver/auth/sasl.go]: Another possible way of doing this is to add an interface named something like `ShortConversationSaslClient`, having the SCRAM SASL adapter implement it, and then doing a type assertion in `ConductSaslConversation` similar to the one for `SaslClientCloser`.

I don't really have a preference for which way this is done, so I'm open to changing it if someone thinks adding another interface would be better.

Comment (jyemin) [reply x/mongo/driver/auth/sasl.go]: The intention for the server API is that the options are mechanism-specific.   So in Java driver implementation we added a method that allows any sasl mechanism to append whatever options it wants.  The default is to append nothing, and for SCRAM mechanisms to append this new option.

File: x/mongo/driver/auth/scram_test.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	}
+	scramSha256ShortPayloads = [][]byte{
+		[]byte("r=rOprNGfwEbeRWgbNEkqO%hvYDpWUa2RaTCAfuxFIlj)hNlF$k0,s=W22ZaJ0SNY7soEsUEjb6gQ==,i=4096"),
+		[]byte("v=6rriTRBi23WpRR/wtup+mMhUZUn/dB5nLTJRsjl95G4="),
+	}
+	scramSha1LongPayloads   = append(scramSha1ShortPayloads, []byte{})
+	scramSha256LongPayloads = append(scramSha256ShortPayloads, []byte{})
+)
+
+func TestSCRAM(t *testing.T) {
```
Comment (jyemin) [inline x/mongo/driver/auth/scram_test.go]: Can this also assert that the driver always appends the skipEmptyExchange option to saslStart?

Does this test assert anything about the client commands being sent?

Comment (divjotarora) [reply x/mongo/driver/auth/scram_test.go]: Added the functionality to parse the command that was sent and an assertion that it has the options we expect.

**Comment by xdg** [comment]
FYI: I'm leaving on vacation until Feb 18, so feel free to proceed without me.

File: x/mongo/driver/auth/sasl.go [hunk 4]
```diff
@@ -28,6 +28,11 @@ type SaslClientCloser interface {
 	Close()
 }
 
+// ExtraOptionsSaslClient is a SaslClient that appens options to the saslStart command.
```
Comment (iwysiu) [inline x/mongo/driver/auth/sasl.go]: appends

File: x/mongo/driver/auth/scram_test.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+			{"scram-sha-256 long conversation", newScramSHA256Authenticator, scramSha256LongPayloads, scramSha256Nonce},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				authenticator, err := tc.createAuthenticatorFn(&Cred{
+					Username: "user",
+					Password: "pencil",
+					Source:   "admin",
+				})
+				assert.Nil(t, err, "newScramSHA1Authenticator error: %v", err)
```
Comment (iwysiu) [inline x/mongo/driver/auth/scram_test.go]: Isn't this sometimes a newScramSHA256Authenticator error?

File: x/mongo/driver/auth/sasl.go (hunk 4 above)
Comment (divjotarora) [reply x/mongo/driver/auth/sasl.go]: Done.

File: x/mongo/driver/auth/scram_test.go (hunk 5 above)
Comment (divjotarora) [reply x/mongo/driver/auth/scram_test.go]: Whoops. This was just a SCRAM-SHA-1 test before I made it table driven. Changed the message to be more generic.

**Pull request merged**